torch/nn/parallel/data_parallel.py

```diff
@@ -52,8 +52,51 @@ class DataParallel(Module):
 
     .. warning::
         Forward and backward hooks defined on :attr:`module` and its submodules
-        won't be invoked anymore, unless the hooks are initialized in the
-        :meth:`forward` method.
+        will be invoked ``len(device_ids)`` times, each with inputs located on
+        a particular device. Particularly, the hooks are only guaranteed to be
+        executed in correct order with respect to operations on corresponding
+        devices. For example, it is not guaranteed that hooks set via
+        :meth:`~torch.nn.Module.register_forward_pre_hook` be executed before
+        `all` ``len(device_ids)`` :meth:`~torch.nn.Module.forward` calls, but
+        that each such hook be executed before the corresponding
+        :meth:`~torch.nn.Module.forward` call of that device.
+
+    .. note::
+        There is a subtlety in using the
+        ``pack sequence -> recurrent network -> unpack sequence`` pattern in a
+        :class:`~torch.nn.Module` wrapped in :class:`~torch.nn.DataParallel`.
+        Input to each the :meth:`forward` on each device will only be part of
+        the entire input. Because the unpack operation
+        :func:`torch.nn.utils.rnn.pad_packed_sequence` by default only pads up
+        to the longest input it sees, i.e., the longest on that particular
+        device, size mismatches will happen when results are gathered together.
+        Therefore, you can instead take advantage of the :attr:`total_length`
+        argument of :func:`~torch.nn.utils.rnn.pad_packed_sequence` to make sure
+        that the :meth:`forward` calls return sequences of same length. For
+        example, you can write::
+
+            from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
+
+            class MyModule(nn.Module):
+                # ... __init__, other methods, etc.
+
+                # padding_input is of shape [B x T x *] (batch_first mode) and contains
+                # the sorted sequences
+                # B is the batch size
+                # T is max sequence length
+                def forward(self, padded_input, input_lengths):
+                    total_length = padded_input.size(1)  # get the max sequence length
+                    packed_input = pack_padded_sequence(padded_input, input_lengths,
+                                                        batch_first=True)
+                    packed_output, _ = self.my_lstm(packed_input)
+                    output, _ = pad_packed_sequence(packed_output, batch_first=True,
+                                                    total_length=total_length)
+                    return output
+
+
+            m = MyModule().cuda()
+            dp_m = nn.DataParallel(m)
```

Comment:
I think that's a bit too much for the API docs. We might want to put it in some kind of FAQ and put a reference to it here